### PR TITLE
Do not suppress `TypeError`s thrown on integrity checks

### PR DESCRIPTION
**Pull request by shvaikalesh**
Current implementation will report false-positive result if `TypeError` is thrown in some `Proxy` traps (`isExtensible`, `ownKeys`, `getOwnPropertyDescriptor`).

### CODE REVIEW

**Comment by meeber** [comment]
LGTM. I wasn't aware that `isSealed` and `isFrozen` was caught by the proxy trap for `isExtensible` but now that I see it, it makes sense.

On a side note, I think we should move `obj === Object(obj)` to a new `isObject` utility function, to make it easier for future contributors to use as well as make it easier to change implementation in the future if we ever need to. (But not in this PR)

**Comment by shvaikalesh** [comment]
@meeber Noted on `isObject` helper. I used `isExtensible` for tests because it is shorter than other traps mentioned: `isFrozen` won't call `ownKeys` unless the object is not extensible. I can tests for other methods, though. But I thought it would be too much code for such an edge case.

**Comment by meeber** [comment]
I think it's good the way you have it. I'd just never used the `isExtensible` trap before, and hadn't thought about it being triggered by `isSealed` and `isFrozen`. But it makes sense. And it's indirectly inferred by the ECMAScript spec.

**Comment by shvaikalesh** [comment]
The [spec](https://tc39.github.io/ecma262/#sec-testintegritylevel) is pretty accurate on which internal methods are called and when. I'have diversified tests a bit: I make `proxy` not extendable and throw `TypeError` in `OwnPropertyKeys` trap for `frozen` and `sealed` tests.

File: lib/chai/core/assertions.js [hunk 1]
```diff
@@ -2232,14 +2227,7 @@ module.exports = function (chai, _) {
     // See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/isSealed
     // The following provides ES6 behavior when a TypeError is thrown under ES5.
 
-    var isSealed;
```
Comment (meeber) [inline lib/chai/core/assertions.js]: Same thing in 2225-2228 with comments update.

File: test/assert.js [hunk 2]
```diff
@@ -1922,6 +1946,20 @@ describe('assert', function () {
...
+            throw new TypeError();
+          }
+        });
+
+        Object.preventExtensions(proxy);
+
+        err(function() {
+          assert[isSealed](proxy);
+        }, { name: 'TypeError' });
+      }
```
Comment (meeber) [inline test/assert.js]: I think the above test is advanced enough that it calls for a short comment explaining why it is the way it is, particularly with regard to the `preventExtensions` and `ownKeys`. A comment will lower the barrier of entry to new contributors and reduce risk of someone submitting a change to it without realizing how it works internally.

File: test/assert.js [hunk 3]
```diff
@@ -1980,6 +2032,20 @@ describe('assert', function () {
...
+            throw new TypeError();
+          }
+        });
+
+        Object.preventExtensions(proxy);
+
+        err(function() {
+          assert[isFrozen](proxy);
+        }, { name: 'TypeError' });
+      }
```
Comment (meeber) [inline test/assert.js]: Same thing with short comment explaining above test.

File: test/assert.js [hunk 4]
```diff
@@ -2014,6 +2080,20 @@ describe('assert', function () {
...
+            throw new TypeError();
+          }
+        });
+
+        Object.preventExtensions(proxy);
+
+        err(function() {
+          assert[isNotFrozen](proxy);
+        }, { name: 'TypeError' });
+      }
```
Comment (meeber) [inline test/assert.js]: Same thing with short comment explaining above test.

File: test/expect.js [hunk 5]
```diff
@@ -2416,6 +2428,20 @@ describe('expect', function () {
...
+          throw new TypeError();
+        }
+      });
+
+      Object.preventExtensions(proxy);
+
+      err(function() {
+        expect(proxy).to.be.sealed;
+      }, { name: 'TypeError' });
+    }
```
Comment (meeber) [inline test/expect.js]: Same thing with short comment explaining above test.

File: test/expect.js [hunk 6]
```diff
@@ -2463,5 +2489,19 @@ describe('expect', function () {
...
+          throw new TypeError();
+        }
+      });
+
+      Object.preventExtensions(proxy);
+
+      err(function() {
+        expect(proxy).to.be.frozen;
+      }, { name: 'TypeError' });
+    }
```
Comment (meeber) [inline test/expect.js]: Same thing with short comment explaining above test.

File: test/should.js [hunk 7]
```diff
@@ -2241,6 +2253,20 @@ describe('should', function() {
...
+          throw new TypeError();
+        }
+      });
+
+      Object.preventExtensions(proxy);
+
+      err(function() {
+        proxy.should.be.sealed;
+      }, { name: 'TypeError' });
+    }
```
Comment (meeber) [inline test/should.js]: Same thing with short comment explaining above test.

**Comment by shvaikalesh** [comment]
@meeber Thanks for reviewing this one, your help is greatly appreciated!

**Comment by meeber** [comment]
My pleasure. Awesome work! This is an important PR for improving our ES6 support, especially since it's part of the `isExtensible` proxy trap's behavior to throw a `TypeError` if it returns an untruthful value. Definitely don't want Chai to swallow that error!

Let's get some more eyes on this before merging: @lucasfcosta @vieiralucas @keithamus

**Comment by lucasfcosta** [comment]
Thanks for the PR @shvaikalesh! Awesome work and thanks for including tests, this will be a great improvement.

LGTM. Also, you seem to be a jedi master in the use of `Object()`. I learned a great lesson with you, I'm gonna use it more often. 😄

**Comment by shvaikalesh** [comment]
@lucasfcosta Thanks, but beware, `stuff === Object(stuff)` used to cause wrong results in older  V8 (Node.js 0.8, Chrome 17-18) JITs. It is safe now.

**Comment by lucasfcosta** [comment]
Does any of you have permissions to restart the build? Apparently it has failed for no reason.

**Comment by meeber** [comment]
Restarted
